File: firmware/scripts/build_all.rb
# frozen_string_literal: true

require 'open3'

BuildUpdatedOnly = ARGV.include?('--updatedOnly')
AbortOnError = ARGV.include?('--abortOnError')

def get_all_project_paths
  Dir.glob('./src/projects/**/rules.mk')
     .map { |path| File.dirname(path) }
     .select { |path| File.exist?(File.join(path, 'project.json')) }
     .map { |path| path.sub('./src/projects/', '') }
end

def get_updated_project_paths
  `git diff --name-only HEAD~`
    .split(/\R/)
    .map { |path| File.dirname(path) }
    .uniq
    .filter { |path| path.start_with?(Projects_dir) }
    .map { |path| path.sub(Projects_dir, '') }
    .filter do |dir|
    ['rules.mk', 'project.json'].all? { |file_name| File.exist?(File.join('src/projects', dir, file_name)) }
  end
end

def make_project_build(project_path)
  command = "make #{project_path}:build"
  _stdout, stderr, status = Open3.capture3(command)
  return { result: :ok } if status == 0

  { result: :ng, error_log: ">#{command}\n#{stderr}" }
end

def check_binary_size(project_path)
  size_command = "make #{project_path}:size"
  size_output_lines = `#{size_command}`
  usage_prog = size_output_lines.match(/^Program.*\(([\d.]+)% Full\)/)[1].to_f
  usage_data = size_output_lines.match(/^Data.*\(([\d.]+)% Full\)/)[1].to_f
  if usage_prog < 100.0 && usage_data < 100.0
    { result: :ok }
  else
    { result: :ng, error_log: "firmware footprint overrun (FLASH: #{usage_prog}, RAM: #{usage_data})" }
  end
end

def build_project(project_path)
  puts "building #{project_path} ..."

  `make #{project_path}:purge`

  res = make_project_build(project_path)
  res = check_binary_size(project_path) if res[:result] == :ok

  if res[:result] == :ok
    print "\e[A\e[K"
    puts "build #{project_path} ... OK"
    true
  else
    puts(res[:error_log])
    if AbortOnError
      warn("abort: failed to build #{project_path}")
      exit(1)
    end
    puts "build #{project_path} ... NG"
    puts
    false
  end
end

def build_projects
  project_paths = BuildUpdatedOnly ? get_updated_project_paths : get_all_project_paths
  `make clean`
  puts "target projects: #{project_paths}"
  results = project_paths.map { |project_path| build_project(project_path) }
  num_success = results.count(true)
  num_total = results.length
  puts "build_stats #{num_success}/#{num_total}" if num_total > 0
  puts 'done'
end

build_projects if __FILE__ == $0

File: firmware/scripts2/buildAll.ts
import * as fs from "fs";
import glob from "glob";
import path from "path";
import { stringifyArray, uniqueArrayItems } from "./helpers";
import { executeCommand } from "./osHelpers";

process.chdir("..");

const puts = console.log;

const FirmwareProjectsDir = "firmware/src/projects";
const BuildUpdatedOnly = process.argv.includes("--updatedOnly");
const AbortOnError = process.argv.includes("--abortOnError");

function getAllProjectPaths() {
  return glob
    .sync("src/projects/**/rules.mk")
    .map((fpath) => path.dirname(fpath))
    .filter((fpath) => fs.existsSync(path.join(fpath, "layout.json")))
    .map((fpath) => path.relative("src/projects", fpath));
}

function getUpdatedProjectPaths() {
  const [gitOutput] = executeCommand(`git diff --name-only HEAD~`);
  const dirs = uniqueArrayItems(
    gitOutput.split(/\r?\n/).map((fpath) => path.dirname(fpath))
  );
  return dirs
    .filter((fpath) => fpath.startsWith(FirmwareProjectsDir))
    .map((fpath) => path.relative(FirmwareProjectsDir, fpath))
    .filter((projectPath) =>
      ["rules.mk", "layout.json"].every((fileName) =>
        fs.existsSync(path.join("src/projects", projectPath, fileName))
      )
    );
}

function makeProjectBuild(projectPath: string) {
  const command = `make ${projectPath}:build`;
  const [_stdout, stderr, status] = executeCommand(command);
  if (status !== 0) {
    throw new Error(`>${command}\n${stderr}`);
  }
}

function checkBinarySize(projectPath: string) {
  const sizeCommand = `make ${projectPath}:size`;
  const [sizeOutputLines] = executeCommand(sizeCommand);
  const usageProg = parseFloat(
    sizeOutputLines.match(/^Program.*\(([\d.]+)% Full\)/m)![1]
  );
  const usageData = parseFloat(
    sizeOutputLines.match(/^Data.*\(([\d.]+)% Full\)/m)![1]
  );
  if (!(usageProg < 100.0 && usageData < 100.0)) {
    throw new Error(
      `firmware footprint overrun (FLASH: ${usageProg}, RAM: ${usageData})`
    );
  }
}

function buildProject(projectPath: string): boolean {
  puts(`build ${projectPath} ...`);
  executeCommand(`make ${projectPath}:purge`);
  try {
    makeProjectBuild(projectPath);
    checkBinarySize(projectPath);
    puts(`build ${projectPath} ... OK`);
    return true;
  } catch (error) {
    puts(error.message);
    if (AbortOnError) {
      console.warn(`abort: failed to build ${projectPath}`);
      process.exit(1);
    }
    puts(`build ${projectPath} ... NG`);
    puts();
    return false;
  }
}

function buildProjects() {
  const projectPaths = BuildUpdatedOnly
    ? getUpdatedProjectPaths()
    : getAllProjectPaths();
  puts(`target projects: ${stringifyArray(projectPaths)}`);

  executeCommand("make clean");
  const results = projectPaths.map(buildProject);

  const numSuccess = results.filter((result) => result).length;
  const numTotal = results.length;
  if (numTotal > 0) {
    puts(`buildStats ${numSuccess}/${numTotal}`);
  }
  puts("done");
}

buildProjects();
